analyses/batch-effects/util/functions.R

```diff
@@ -0,0 +1,361 @@
+  id_batch_histology = as_tibble(id_batch_histology)
+  
+  # convert id_batch_histology from a matrix to a tibble
+  id_batch_histology = as_tibble(id_batch_histology)
+  
+  return(id_batch_histology)
+}
+
+# make two PCA plots before and after applying combat
+make_before_and_after_combat_plots = function(data_t, batch, report_name, run_combat) {
+  
+  data_t$batch = batch_prep(data_t)
+  
+  #make a matrix representing just the gene expression data (this gets plugged into combat)
+  poly1matrix = t(data_t[,4:ncol(data_t)])
+  
+  pca = prcomp(log2(data_t[,4:ncol(data_t)]+ 1))
+  
+  tb = as_tibble(pca$x)
+  
+  p1 = make_plot(tb, data_t, report_name, "histology.png")
+  
+  if (run_combat == FALSE){
+    print("Combat set to not run on this data set")
+    return()
+  }
+  
+  plot_list = tryCatch({
+    # filter out genes without variance
+    var_zero_genes = poly1matrix[which((apply(poly1matrix, 1, var)==0)),]
+    poly1matrix = poly1matrix[which((apply(poly1matrix, 1, var)>0)),]
+    print(sum(var_zero_genes))
+    
+    poly2matrix <- log2(poly1matrix + 1)
+    
+    
+    combat_matrix = ComBat(dat=poly2matrix, batch=batch)
+    
+    pca = prcomp(t(combat_matrix))
+    
+    tb = as_tibble(pca$x)
+    
+    p2 = make_plot(tb, data_t, report_name, "batch-adjusted-histology.png")
+    return(list(p1, p2, data_t))
+  }, error=function(cond){
+    message("ComBat failed! It's likely that you are trying to adjust for batches with just 1 batch")
+    # message(cond)
+    return(list(p1, NULL, data_t))
+  })
+  
+  return(plot_list)
+  
+}
+
+# only include three most common histologies
+histology_filtering = function(data_t){
+  # create a list of the names of the three most common histologies
+  short_histo = names(sort(table((data_t$histology)), decreasing = TRUE)[1:3])
+  
+  # filter the data to only include the top 3
+  data_t = filter(data_t, histology %in% short_histo)
+  
+  return(data_t)
+}
+
+# root functions that calls sub functions to make all of the graphs
+make_histology_pca_plots = function(df, id_batch_histology, gene_id, report_name, file_name, run_combat=TRUE){
+  
+  #handle file outputs
+  report_name = handle_file_outputs(report_name)
+ 
+  # clean the data see tible_combat_prep comments
+  data_t = tibble_combat_prep(df, gene_id, id_batch_histology)
+  
+  # only include three most common histologies
+  data_t = histology_filtering(data_t)
+  
+  # see batch_prep comments
+  batch = batch_prep(data_t)
+  
+  print("DATA IS READY") 
+  
+  #Make before and after PCA plots
+  return(make_before_and_after_combat_plots(data_t, batch, report_name, run_combat))
+
+}
+
+# root functions that calls sub functions to combat and batchQC looking for batch affects
+run_batchQC = function(df, id_batch, gene_id, report_name, file_name, run_combat){
+  
+  # create output directory (if necessary)
+  dir.create(file.path(analysis_dir, "results"), showWarnings = FALSE)
```

Comment:
This `util/functions.R` file could be sourced from anywhere, which means that where it gets sources may or may not have `analysis` directories in the global environment. You could pass `analysis_dir` as an argument here (or maybe plan to have `file_name` contain the full path) or you can include this in a code block in the notebook you're intending to use it.

Reply:
I'll add analysis_dir as a parameter.